handlers/arcgis.go

```diff
@@ -0,0 +1,367 @@
+			Capabilities:      "Map",
+		}
+
+		out := map[string]interface{}{
+			"layers": layers,
+		}
+
+		bytes, err := json.Marshal(out)
+		if err != nil {
+			return http.StatusInternalServerError, fmt.Errorf("cannot marshal ArcGIS layer info JSON: %v", err)
+		}
+
+		return http.StatusOK, wrapJSONP(w, r, bytes)
+	}
+}
+
+func (s *ServiceSet) arcgisLegend(id string, db *mbtiles.DB) handlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) (int, error) {
+
+		metadata, err := db.ReadMetadata()
+		if err != nil {
+			return http.StatusInternalServerError, fmt.Errorf("Could not read metadata for tileset %v", id)
+		}
+
+		// TODO: pull the legend from ArcGIS specific metadata tables
+		var elements [0]interface{}
+		var layers [1]map[string]interface{}
+
+		layers[0] = map[string]interface{}{
+			"layerId":   0,
+			"layerName": toString(metadata["name"]),
+			"layerType": "",
+			"minScale":  0,
+			"maxScale":  0,
+			"legend":    elements,
+		}
+
+		out := map[string]interface{}{
+			"layers": layers,
+		}
+
+		bytes, err := json.Marshal(out)
+		if err != nil {
+			return http.StatusInternalServerError, fmt.Errorf("cannot marshal ArcGIS legend info JSON: %v", err)
+		}
+		return http.StatusOK, wrapJSONP(w, r, bytes)
+	}
+}
+
+func (s *ServiceSet) arcgisTiles(db *mbtiles.DB) handlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) (int, error) {
+		// split path components to extract tile coordinates x, y and z
+		pcs := strings.Split(r.URL.Path[1:], "/")
+		// strip off /arcgis/rest/services/ and then
+		// we should have at least <id> , "MapServer", "tiles", <z>, <y>, <x>
+		l := len(pcs)
+		if l < 6 || pcs[5] == "" {
+			return http.StatusBadRequest, fmt.Errorf("requested path is too short")
+		}
+		z, y, x := pcs[l-3], pcs[l-2], pcs[l-1]
+		tc, _, err := tileCoordFromString(z, x, y)
+		if err != nil {
+			return http.StatusBadRequest, err
+		}
+
+		var data []byte
+		err = db.ReadTile(tc.z, tc.x, tc.y, &data)
+
+		if err != nil {
+			// augment error info
+			t := "tile"
+			err = fmt.Errorf("cannot fetch %s from DB for z=%d, x=%d, y=%d: %v", t, tc.z, tc.x, tc.y, err)
+			return http.StatusInternalServerError, err
+		}
+
+		if data == nil || len(data) <= 1 {
+			// Return blank PNG for all image types
+			w.Header().Set("Content-Type", "image/png")
+			_, err = w.Write(BlankPNG())
+		} else {
+			w.Header().Set("Content-Type", db.ContentType())
+			_, err = w.Write(data)
+		}
+
+		return http.StatusOK, err
+	}
+}
+
+func geoBoundsToWMExtent(bounds []float32) arcGISExtent {
+	xmin, ymin := geoToMercator(float64(bounds[0]), float64(bounds[1]))
+	xmax, ymax := geoToMercator(float64(bounds[2]), float64(bounds[3]))
+	return arcGISExtent{
+		Xmin:             xmin,
+		Ymin:             ymin,
+		Xmax:             xmax,
+		Ymax:             ymax,
+		SpatialReference: webMercatorSR,
+	}
+}
+
+func calcScaleResolution(zoomLevel int, dpi int) (float64, float64) {
+	resolution := 156543.033928 / math.Pow(2, float64(zoomLevel))
```

Comment:
For readability I'd suggest using a `const` block for declaring these constants.  

If you replace `math.Pow(...)` with `float64(1<<uint(zoomLevel))`, it can be a constant, too. Also I'd consider using an `uint` for the `zoomLevel`, whereever possible.

Reply:
```go
const(
    earthRadius = 6378137.0
    earthCircumference = math.Pi * earthRadius
    initialResolution =  2 * earthCircumference / 256 
) 
resolution := initialResolution / float(1 << zoomLevel)
```

The constants will be pre-computed as much as possible by the compiler.

Reply:
@fawick I don't follow how `float64(1<<uint(zoomLevel))` would be better as a const in this function than just used directly in the calculation.  Can you please clarify?

The rest of your suggestions here make sense, and I moved the other constants you listed here up as unexported package constants.

Reply:
Of course, you're right! My remark that it could be a constant is indeed wrong, when I wrote it initially I overlooked that the expression depends on the parameter `zoomLevel` and failed to correct that. Sorry for the confusion! 
What I wanted to express first and foremost is that in the special case of powers of two, `math.Pow` is a relative expensive numeric function call in comparison to the single CPU instruction needed for the arithmetic left shift.